allennlp/tests/modules/seq2seq_decoders/auto_regressive_seq_decoder_test.py

```diff
@@ -0,0 +1,131 @@
+        )
+
+        encoded_state = torch.rand(batch_size, time_steps, decoder_inout_dim)
+        source_mask = torch.ones(batch_size, time_steps).long()
+        target_tokens = {"tokens": torch.ones(batch_size, time_steps).long()}
+        source_mask[0, 1:] = 0
+        encoder_out = {"source_mask": source_mask, "encoder_outputs": encoded_state}
+
+        assert auto_regressive_seq_decoder.forward(encoder_out) == {}
+        loss = auto_regressive_seq_decoder.forward(encoder_out, target_tokens)["loss"]
+        assert loss.shape == torch.Size([]) and loss.requires_grad
+        auto_regressive_seq_decoder.eval()
+        assert "predictions" in auto_regressive_seq_decoder.forward(encoder_out)
+
+    def test_auto_regressive_seq_decoder_indices_to_tokens(self):
+        decoder_inout_dim = 4
+        vocab, decoder_net = create_vocab_and_decoder_net(decoder_inout_dim)
+
+        auto_regressive_seq_decoder = AutoRegressiveSeqDecoder(
+            vocab, decoder_net, 10, Embedding(vocab.get_vocab_size(), decoder_inout_dim)
+        )
+
+        predictions = torch.tensor([[3, 2, 5, 0, 0], [2, 2, 3, 5, 0]])
+
+        tokens_ground_truth = [["B", "A"], ["A", "A", "B"]]
+        predicted_tokens = auto_regressive_seq_decoder.indices_to_tokens(predictions.numpy())
+        assert predicted_tokens == tokens_ground_truth
+
+    def test_auto_regressive_seq_decoder_post_process(self):
+        decoder_inout_dim = 4
+        vocab, decoder_net = create_vocab_and_decoder_net(decoder_inout_dim)
+
+        auto_regressive_seq_decoder = AutoRegressiveSeqDecoder(
+            vocab, decoder_net, 10, Embedding(vocab.get_vocab_size(), decoder_inout_dim)
+        )
+
+        predictions = torch.tensor([[3, 2, 5, 0, 0], [2, 2, 3, 5, 0]])
+
+        tokens_ground_truth = [["B", "A"], ["A", "A", "B"]]
+
+        output_dict = {"predictions": predictions}
+        predicted_tokens = auto_regressive_seq_decoder.post_process(output_dict)["predicted_tokens"]
+        assert predicted_tokens == tokens_ground_truth
+
+    def test_auto_regressive_seq_decoder_tensor_and_token_based_metric(self):
+        prepare_environment(Params({}))
+
+        batch_size, time_steps, decoder_inout_dim = 2, 3, 4
+        vocab, decoder_net = create_vocab_and_decoder_net(decoder_inout_dim)
+
+        auto_regressive_seq_decoder = AutoRegressiveSeqDecoder(
+            vocab,
+            decoder_net,
+            10,
+            Embedding(vocab.get_vocab_size(), decoder_inout_dim),
+            tensor_based_metric=BLEU(),
+            token_based_metric=CustomSquadEmAndF1(),
+        ).eval()
+
+        encoded_state = torch.randn(batch_size, time_steps, decoder_inout_dim)
+        source_mask = torch.ones(batch_size, time_steps).long()
+        target_tokens = {"tokens": torch.ones(batch_size, time_steps).long()}
+        source_mask[0, 1:] = 0
+        encoder_out = {"source_mask": source_mask, "encoder_outputs": encoded_state}
```

Comment:
Could you clarify why the asserted metric are fixed even though the encoded state is random? Looks great otherwise! Thanks for the tests!

Reply:
The encoder state is not random. `prepare_environment(Params({}))` is setting all the seeds to a fixed value so `torch.randn` is returning always the same value and that is why the test passes when checking the BLEU score.

Reply:
Ah, right. Thanks! Could you add that as a comment?

Reply:
Done! ;)